src/firmware/main/app/primitives/primitive_manager.c

```diff
@@ -0,0 +1,185 @@
+#include "primitive_manager.h"
+
+// There are different semaphore implementations depending on if we're on a x86 or arm
+// system, and we use typdefs here to switch between them
+#ifdef __arm__
+#include <FreeRTOS.h>
+#include <semphr.h>
+#elif __unix__
+#include <pthread.h>
+#else
+#error "Could not determine what CPU this is being compiled for."
+#endif
+
+#include <assert.h>
+#include <stdint.h>
+#include <stdio.h>
+
+#include "firmware/main/app/primitives/catch_primitive.h"
+#include "firmware/main/app/primitives/direct_velocity_primitive.h"
+#include "firmware/main/app/primitives/direct_wheels_primitive.h"
+#include "firmware/main/app/primitives/dribble_primitive.h"
+#include "firmware/main/app/primitives/imu_test_primitive.h"
+#include "firmware/main/app/primitives/move_primitive.h"
+#include "firmware/main/app/primitives/pivot_primitive.h"
+#include "firmware/main/app/primitives/primitive.h"
+#include "firmware/main/app/primitives/shoot_primitive.h"
+#include "firmware/main/app/primitives/spin_primitive.h"
+#include "firmware/main/app/primitives/stop_primitive.h"
+
+struct PrimitiveManager
+{
+// The mutex that prevents multiple entries into the same primitive at the same time.
+#ifdef __arm__
+    SemaphoreHandle_t primitive_mutex;
+#elif __unix__
+    pthread_mutex_t primitive_mutex;
+#endif
+
+    // TODO: rename, this is really a pointer to the _functions_ of the current
```

Comment:
TODO: Rename what exactly?

Reply:
Ah, so I renamed `current_primitive` to `current_primitive_functions` because I want it to be clear that this member does not hold the _state_ of the primitive.